core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2615,10 +2615,13 @@ public void testHandleDocumentFailure() throws Exception {
                 }
                 Engine.IndexResult indexResult = engine.index(indexForDoc(doc1));
                 assertNotNull(indexResult.getFailure());
-
+                // document failures should be recorded in translog
+                assertNotNull(indexResult.getTranslogLocation());
```

Comment:
there is already a test for engine `noOp` operation here, which the replica uses to record failures. I added a unit test (`testNoOpReplicationOnPrimaryDocumentFailure`) in `TransportShardBulkActionTests` to verify the noOp parameters when a primary document failure is seen